File: android-native/app/src/main/java/com/example/kpkn/services/competition/CompetitionReminderManager.kt
package com.example.kpkn.services.competition

import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.example.kpkn.data.models.CompetitionRecord
import com.example.kpkn.navigation.KpknDeepLinks
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId

class CompetitionReminderManager(context: Context) {

    private val appCtx = context.applicationContext
    private val alarmManager = appCtx.getSystemService(Context.ALARM_SERVICE) as AlarmManager

    fun createChannels() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val channel = NotificationChannel(
            CHANNEL_COMPETITION,
            "Competiciones",
            NotificationManager.IMPORTANCE_DEFAULT,
        ).apply {
            description = "Recordatorios de eventos competitivos"
            enableVibration(true)
        }
        val manager = appCtx.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.createNotificationChannel(channel)
    }

    fun schedule(record: CompetitionRecord) {
        cancel(record.id)
        val eventAt = record.eventDate?.toEventDateTime(record.startTime) ?: return
        scheduleIfEnabled(record, TYPE_WEEK, eventAt.minusDays(7), record.reminderOneWeekEnabled)
        scheduleIfEnabled(record, TYPE_48H, eventAt.minusHours(48), record.reminder48hEnabled)
        scheduleIfEnabled(record, TYPE_START, eventAt, record.reminderStartEnabled)
    }

    fun cancel(recordId: String) {
        listOf(TYPE_WEEK, TYPE_48H, TYPE_START).forEach { type ->
            alarmManager.cancel(pendingIntent(recordId, type))
        }
    }

    private fun scheduleIfEnabled(record: CompetitionRecord, type: String, triggerAt: LocalDateTime, enabled: Boolean) {
        if (!enabled) return
        val triggerMs = triggerAt.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
        if (triggerMs <= System.currentTimeMillis()) return
        val pendingIntent = pendingIntent(record.id, type, record.title, record.eventDate.orEmpty())
        runCatching {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && alarmManager.canScheduleExactAlarms()) {
                alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerMs, pendingIntent)
            } else {
                alarmManager.set(AlarmManager.RTC_WAKEUP, triggerMs, pendingIntent)
            }
        }
    }

    private fun pendingIntent(
        recordId: String,
        type: String,
        title: String = "",
        date: String = "",
    ): PendingIntent {
        val requestCode = (recordId + type).hashCode()
        val intent = Intent(appCtx, CompetitionReminderReceiver::class.java).apply {
            putExtra(EXTRA_RECORD_ID, recordId)
            putExtra(EXTRA_TYPE, type)
            putExtra(EXTRA_TITLE, title)
            putExtra(EXTRA_DATE, date)
        }
        return PendingIntent.getBroadcast(
            appCtx,
            requestCode,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    companion object {
        const val CHANNEL_COMPETITION = "competition_reminders"
        const val EXTRA_RECORD_ID = "competition_record_id"
        const val EXTRA_TYPE = "competition_reminder_type"
        const val EXTRA_TITLE = "competition_title"
        const val EXTRA_DATE = "competition_date"
        const val TYPE_WEEK = "week"
        const val TYPE_48H = "48h"
        const val TYPE_START = "start"
    }
}

class CompetitionReminderReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (!hasPermission(context)) return
        CompetitionReminderManager(context).createChannels()

        val title = intent.getStringExtra(CompetitionReminderManager.EXTRA_TITLE)
            ?.takeIf { it.isNotBlank() }
            ?: "Competición"
        val date = intent.getStringExtra(CompetitionReminderManager.EXTRA_DATE).orEmpty()
        val type = intent.getStringExtra(CompetitionReminderManager.EXTRA_TYPE).orEmpty()
        val body = when (type) {
            CompetitionReminderManager.TYPE_WEEK -> "Queda una semana para $title."
            CompetitionReminderManager.TYPE_48H -> "Quedan 48 horas para $title."
            CompetitionReminderManager.TYPE_START -> "$title comienza ahora."
            else -> "Revisa tu registro de competición."
        } + if (date.isNotBlank()) " Fecha: $date." else ""

        val notification = NotificationCompat.Builder(context, CompetitionReminderManager.CHANNEL_COMPETITION)
            .setSmallIcon(com.example.kpkn.R.mipmap.ic_launcher)
            .setContentTitle("KPKN · Competición")
            .setContentText(body)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
            .setContentIntent(
                KpknDeepLinks.pendingActivityIntent(
                    context = context,
                    requestCode = 0,
                    path = "competitions",
                )
            )
            .build()

        runCatching {
            NotificationManagerCompat.from(context).notify((title + date + type).hashCode(), notification)
        }
    }

    private fun hasPermission(context: Context): Boolean =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ContextCompat.checkSelfPermission(context, android.Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }
}

private fun String.toEventDateTime(startTime: String?): LocalDateTime? {
    val date = runCatching { LocalDate.parse(this) }.getOrNull() ?: return null
    val time = startTime?.let { runCatching { LocalTime.parse(it) }.getOrNull() } ?: LocalTime.of(9, 0)
    return LocalDateTime.of(date, time)
}

File: ios-native/KPKNFit/KPKNFit/Services/Competition/CompetitionReminderManager.swift
import Foundation
import UIKit
import UserNotifications
import os.log

public enum CompetitionReminderManager {

    public static let channelCompetition = "competition_reminders"
    public static let extraRecordId = "competition_record_id"
    public static let extraType = "competition_reminder_type"
    public static let extraTitle = "competition_title"
    public static let extraDate = "competition_date"

    public static let typeWeek = "week"
    public static let type48h = "48h"
    public static let typeStart = "start"
    public static let typePostResult = "post_result"

    public static func registerCategories() {
        let category = UNNotificationCategory(
            identifier: channelCompetition,
            actions: [],
            intentIdentifiers: [],
            options: []
        )
        UNUserNotificationCenter.current().setNotificationCategories([category])
    }

    public static func schedule(record: CompetitionRecord) {
        cancel(recordId: record.id)
        guard let eventAt = record.eventDate?.toEventDateTime(startTime: record.startTime) else { return }

        scheduleIfEnabled(record: record, type: typeWeek, triggerAt: eventAt.addingTimeInterval(-7 * 24 * 3600), enabled: record.reminderOneWeekEnabled)
        scheduleIfEnabled(record: record, type: type48h, triggerAt: eventAt.addingTimeInterval(-48 * 3600), enabled: record.reminder48hEnabled)
        scheduleIfEnabled(record: record, type: typeStart, triggerAt: eventAt, enabled: record.reminderStartEnabled)

        if !(record.startTime ?? "").isEmpty {
            scheduleIfEnabled(record: record, type: typePostResult, triggerAt: eventAt.addingTimeInterval(5 * 3600), enabled: true)
        }
    }

    public static func cancel(recordId: String) {
        let identifiers = [typeWeek, type48h, typeStart, typePostResult].map { "competition_\(recordId)_\($0)" }
        UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: identifiers)
    }

    private static func scheduleIfEnabled(record: CompetitionRecord, type: String, triggerAt: Date, enabled: Bool) {
        guard enabled else { return }
        guard triggerAt > Date() else { return }

        let content = UNMutableNotificationContent()
        content.title = "KPKN · Competición"
        content.sound = .default
        content.userInfo = [
            extraRecordId: record.id,
            extraType: type,
            extraTitle: record.title,
            extraDate: record.eventDate ?? ""
        ]

        let body: String = {
            switch type {
            case typeWeek: return "Queda una semana para \(record.title)."
            case type48h: return "Quedan 48 horas para \(record.title)."
            case typeStart: return "\(record.title) comienza ahora."
            case typePostResult: return "¿Cómo te fue en la competición?"
            default: return "Revisa tu registro de competición."
            }
        }()
        if let date = record.eventDate, !date.isEmpty {
            content.body = body + " Fecha: \(date)."
        } else {
            content.body = body
        }

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: triggerAt.timeIntervalSinceNow, repeats: false)
        let request = UNNotificationRequest(
            identifier: "competition_\(record.id)_\(type)",
            content: content,
            trigger: trigger
        )

        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                os_log("Failed to schedule competition reminder: %{public}@",
                       log: .default, type: .error, error.localizedDescription)
            }
        }
    }

    public static func handleNotification(userInfo: [AnyHashable: Any]) {
        let recordId = userInfo[extraRecordId] as? String ?? ""
        let title = (userInfo[extraTitle] as? String)?.nilIfBlank ?? "Competición"
        let date = userInfo[extraDate] as? String ?? ""
        let type = userInfo[extraType] as? String ?? ""

        if type == typePostResult && !recordId.isEmpty {
            Task {
                let entity = await KpknDatabase.instance().competitionRecordDao.getById(id: recordId)
                if entity?.status != CompetitionRecordStatus.PLANNED.rawValue { return }
                showNotification(title: title, body: makeBody(type: type, title: title, date: date), type: type, recordId: recordId)
            }
        } else {
            showNotification(title: title, body: makeBody(type: type, title: title, date: date), type: type, recordId: recordId)
        }
    }

    private static func makeBody(type: String, title: String, date: String) -> String {
        let base: String = {
            switch type {
            case typeWeek: return "Queda una semana para \(title)."
            case type48h: return "Quedan 48 horas para \(title)."
            case typeStart: return "\(title) comienza ahora."
            case typePostResult: return "¿Cómo te fue en la competición?"
            default: return "Revisa tu registro de competición."
            }
        }()
        return date.isEmpty ? base : base + " Fecha: \(date)."
    }

    private static func showNotification(title: String, body: String, type: String, recordId: String) {
        let content = UNMutableNotificationContent()
        content.title = "KPKN · Competición"
        content.body = body
        content.sound = .default
        content.userInfo = [extraRecordId: recordId, extraType: type]

        let request = UNNotificationRequest(
            identifier: "competition_delivered_\(recordId)_\(type)",
            content: content,
            trigger: nil
        )
        UNUserNotificationCenter.current().add(request) { _ in }
    }
}

private extension String {
    var nilIfBlank: String? {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
    }

    func toEventDateTime(startTime: String?) -> Date? {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: self) else { return nil }

        let time: Date
        if let st = startTime {
            let tf = DateFormatter()
            tf.locale = Locale(identifier: "en_US_POSIX")
            tf.dateFormat = "HH:mm"
            if let parsed = tf.date(from: st) {
                time = parsed
            } else {
                time = Calendar.current.date(from: DateComponents(hour: 9, minute: 0))!
            }
        } else {
            time = Calendar.current.date(from: DateComponents(hour: 9, minute: 0))!
        }

        let calendar = Calendar.current
        let dateComponents = calendar.dateComponents([.year, .month, .day], from: date)
        let timeComponents = calendar.dateComponents([.hour, .minute], from: time)
        return calendar.date(from: DateComponents(
            year: dateComponents.year,
            month: dateComponents.month,
            day: dateComponents.day,
            hour: timeComponents.hour,
            minute: timeComponents.minute
        ))
    }
}
